### PR TITLE
Button/ButtonBase: add new props - as and asProps

### PR DESCRIPTION
**Pull request by rogervera**
## Allow passing arbitrary component override to Button component

Description:
There are cases where we need to use `redux-first-router-link` `<Link/>` components that are styled as Buttons.
See: https://github.com/codecademy-engineering/Codecademy/pull/14891#issuecomment-546399761

Spoke with @jakemhiller, and he was thinking of doing this already. Will allow some neat flexibility.

---

## Merging your changes

When you are ready to merge to master, use the "squash and merge" button in GitHub, and follow the [commit message guide](/README.md#commit-message-guide) to write your commit message.

### CODE REVIEW

File: packages/gamut/src/ButtonBase/index.tsx [hunk 1]
```diff
@@ -24,13 +35,24 @@ export type ButtonBaseProps = (
...
 
+  if (As) {
+    <As
+      {...asProps}
+      data-btn
+      {...propsToTransfer}
+      className={classes}
+      onClick={onClick}
+    />;
+  }
```
Comment (JoshuaKGoldberg) [inline packages/gamut/src/ButtonBase/index.tsx]: This looks like a duplicate list of props except for `href`. Could you duplicate in some way?

E.g. `const [Component, props] = someConditionOrFunction`?

File: packages/gamut/src/ButtonBase/index.tsx [hunk 2]
```diff
@@ -3,6 +3,7 @@ import PropTypes from 'prop-types';
 import cx from 'classnames';
 import omitProps from '../utils/omitProps';
 import styles from './styles.scss';
+import { ChildComponentDescriptor } from '../typings/react';
```
Comment (JoshuaKGoldberg) [inline packages/gamut/src/ButtonBase/index.tsx]: Oh hey I think I learned about `React.ComponentType` _after_ adding this... would you mind trying that out instead for this new work? It's used in the monolith in a couple of places.

Bonus points would be if you replace it elsewhere, but that should probably be a separate PR...

File: packages/gamut/src/Button/index.tsx [hunk 3]
```diff
@@ -33,6 +34,16 @@ const propTypes = {
 };
 
 export type ButtonProps = HTMLAttributes<HTMLButtonElement> & {
```
Comment (JoshuaKGoldberg) [inline packages/gamut/src/Button/index.tsx]: 🤔 should there be some common props type these both extend from? Since you're adding the same props with verbose docs to both of them...

File: packages/gamut/src/ButtonBase/index.tsx (hunk 1 above)
Comment (rogervera) [reply packages/gamut/src/ButtonBase/index.tsx]: @JoshuaKGoldberg Something like this?

Comment (JoshuaKGoldberg) [reply packages/gamut/src/ButtonBase/index.tsx]: Yeah that LGTM!

File: packages/gamut/src/ButtonBase/index.tsx (hunk 2 above)
Comment (rogervera) [reply packages/gamut/src/ButtonBase/index.tsx]: Will look at this in a separate PR. Thanks Josh!

File: packages/gamut/src/ButtonBase/index.tsx [hunk 4]
```diff
@@ -24,22 +35,29 @@ export type ButtonBaseProps = (
...
-  );
+  const defaultProps = {
+    ...propsToTransfer,
+    className: classes,
+    onClick: onClick,
+    'data-btn': true,
+  };
+
+  if (href) {
+    return <a {...defaultProps} href={href} />;
```
Comment (rogervera) [inline packages/gamut/src/ButtonBase/index.tsx]: @JoshuaKGoldberg Is there a way to disable the `error  Anchors must have content and the content must be accessible by a screen reader  jsx-a11y/anchor-has-content` for this line? 
Now that we're explicitly using an `<a>` this is complaining. However, there **are** children present.

https://circleci.com/gh/Codecademy/client-modules/13902?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link

File: packages/gamut/src/Button/index.tsx (hunk 3 above)
Comment (rogervera) [reply packages/gamut/src/Button/index.tsx]: Will put this as a nice to have for now...Right now I'd say it's not bad to be WET
https://dev.to/wuz/stop-trying-to-be-so-dry-instead-write-everything-twice-wet-5g33

File: packages/gamut/src/ButtonBase/index.tsx (hunk 4 above)
Comment (JoshuaKGoldberg) [reply packages/gamut/src/ButtonBase/index.tsx]: Seeeeeems like a bug in the lint rule? Does it still happen if you try out #528?

If not, you can disable the lint rule with an `eslint-disable` comment, ideally with a quick explanation of why in another comment.

Comment (rogervera) [reply packages/gamut/src/ButtonBase/index.tsx]: Nope. Will add `es-lint-disable`

File: packages/gamut/src/ButtonBase/index.tsx [hunk 5]
```diff
@@ -24,22 +35,31 @@ export type ButtonBaseProps = (
...
+
+  if (href) {
+    // Anchor tag receives children content from propsToTransfer
+    // eslint-disable-next-line jsx-a11y/anchor-has-content
+    return <a {...defaultProps} href={href} />;
+  }
+
+  if (As) {
+    return <As {...defaultProps} {...asProps} />;
+  }
```
Comment (jakemhiller) [inline packages/gamut/src/ButtonBase/index.tsx]: What do you think about moving the `As` block above `href`? As a consumer of this component I think I would expect whatever I pass in `As` to be respected regardless of the other props.

File: packages/gamut/src/ButtonBase/__tests__/Button-test.tsx [hunk 6]
```diff
@@ -8,12 +8,17 @@ describe('<ButtonBase>', () => {
...
+  it('uses an <a> tag when you pass in an href and the As prop is omitted', () => {
     const wrapper = shallow(<ButtonBase href="/awesome" />);
     expect(wrapper.find('a').length).toEqual(1);
   });
 
-  it('uses a button tag when you omit an href', () => {
+  it('uses a <button> tag when you omit an href and the As prop', () => {
     const wrapper = shallow(<ButtonBase />);
     expect(wrapper.find('button').length).toEqual(1);
   });
```
Comment (jakemhiller) [inline packages/gamut/src/ButtonBase/__tests__/Button-test.tsx]: Might want to add a small test for `asProps` as well

Comment (rogervera) [reply packages/gamut/src/ButtonBase/__tests__/Button-test.tsx]: @jakemhiller Thoughts on this?
https://github.com/Codecademy/client-modules/pull/534/commits/ae01df7b976194012c8cb2723115d0db86254275

File: packages/gamut/src/ButtonBase/index.tsx [hunk 7]
```diff
@@ -24,22 +35,31 @@ export type ButtonBaseProps = (
 
 export const ButtonBase = (props: ButtonBaseProps) => {
   const { href, className, link, onClick } = props;
-  const propsToTransfer = omitProps(propTypes, props);
+  const { as: As, asProps = {}, ...restOfProps } = props;
```
Comment (jakemhiller) [inline packages/gamut/src/ButtonBase/index.tsx]: Can we combine these lines pulling items off of props?
